plutus-scb/src/Plutus/SCB/Core/ContractInstance.hs

```diff
@@ -0,0 +1,484 @@
+            let response = either WriteTxFailed (WriteTxSuccess . Ledger.txId) r
+            logInfo . render $ "processWriteTxRequest result:" <+> pretty response
+            sendContractInboxMessage (WriteTxResponse response) i it
+    in fold <$> traverse go txns
+
+-- | A wrapper around the NodeAPI function that returns some more
+-- useful evidence of the work done.
+submitTx :: (Member WalletEffect effs) => Ledger.Tx -> Eff effs Ledger.Tx
+submitTx tx = submitTxn tx >> pure tx
+
+processNextTxAtRequests ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member Log effs
+    )
+    => Eff effs ()
+processNextTxAtRequests = do
+    logInfo "processNextTxAtRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.nextTxAtRequests @t)
+    traverse_ (uncurry $ processNextTxAtRequest @t) (Map.toList state)
+    logInfo "processNextTxAtRequests end"
+
+processNextTxAtRequest ::
+    forall t effs.
+    ( Member Log effs )
+    => ContractInstanceId
+    -> (ContractIteration, AddressSet)
+    -> Eff effs [ChainEvent t]
+processNextTxAtRequest i (_, AddressSet addr) = do
+    let go address = do
+            logDebug . render $ "processNextTxAtRequest" <+> pretty i <+> pretty address
+            pure []
+    fold <$> traverse go (Set.toList addr)
+
+processTxConfirmedRequests ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member ChainIndexEffect effs
+    , Member Log effs
+    )
+    => Eff effs ()
+processTxConfirmedRequests = do
+    logInfo "processTxConfirmedRequests start"
+    state <- Query.contractStates <$> runGlobalQuery (Query.awaitTxConfirmedRequests @t)
+    traverse_ (uncurry $ processTxConfirmedRequest @t) (Map.toList state)
+    logInfo "processTxConfirmedRequests end"
+
+processTxConfirmedRequest ::
+    forall t effs.
+    ( Member (EventLogEffect (ChainEvent t)) effs
+    , Member ChainIndexEffect effs
+    , Member Log effs
+    )
+    => ContractInstanceId
+    -> (ContractIteration, TxIdSet)
+    -> Eff effs [ChainEvent t]
+processTxConfirmedRequest i (it, TxIdSet txids) =
+    let go txid = do
+            logDebug . render $ "processTxConfirmedRequest" <+> pretty i <+> pretty txid
+            confirmed <- transactionConfirmed txid
+            if confirmed
+                then sendContractInboxMessage (AwaitTxConfirmedResponse $ TxConfirmed txid) i it
+                else pure []
+    in fold <$> traverse go (Set.toList txids)
+
+-- | Look at the outboxes of all contract instances and process them.
+processAllContractOutboxes ::
+    forall t effs.
+    ( Member Log effs
+    , Member (EventLogEffect (ChainEvent t)) effs
+    , Member WalletEffect effs
+    , Member ChainIndexEffect effs
+    , Member SigningProcessEffect effs
+    )
+    => Eff effs ()
+processAllContractOutboxes = do
+    logInfo "processAllContractOutboxes start"
+    -- TODO: maybe this should be interleaved with 'processAllContractInboxes'
```

Comment:
Yeah, I think we discussed this somewhere before. I'm not really sure what the right strategy is. But I think perhaps contracts need explicit outboxes, so we process their outgoing messages in the order they were sent, rather than all the ones of a particular type at once.

Reply:
yeah that's a general question of how things should be organised, by contract instance or by component? Maybe having a dedicated message processor for each instance would also make sense. I don't know.